syntax = "proto3";

option csharp_namespace = "GrpcCustomerServer";

package customer;

service CustomerService {
  rpc GetCustomer (CustomerRequest) returns (CustomerResponse);
  rpc GetCustomerList (CustomerFilterRequest) returns (stream CustomerResponse);
  rpc GetAllCustomers (EmptyRequest) returns (stream CustomerResponse);
  rpc UploadCustomers (stream CustomerResponse) returns (UploadResponse);
  rpc ProcessCustomerStream (stream CustomerResponse) returns (stream ValidationResponse);
}

message CustomerResponse {
  int32 id = 1;
  string name = 2;
  string email = 3;
  string registration_date = 4;
}

message ValidationResponse {
  int32 id = 1;
  string name = 2;
  bool is_valid = 3;
  string message = 4;
}

message CustomerRequest {
  int32 id = 1;
}

message CustomerFilterRequest {
  string keyword = 1;
}

message EmptyRequest {
}

message UploadResponse {
  int32 total_uploaded = 1;
  string message = 2;
}